pulpcore/pulpcore/tasking/services/storage.py

```diff
@@ -1,152 +1,102 @@
-            raise
-
-
-def _rmtree_fix_permissions(directory_path):
-    """Delete an entire directory tree in path.
-
-    Recursively remove a directory. If permissions on the directory or it's contents
-    block removal, attempt to fix the permissions to allow removal and attempt the removal
-    again.
-
-    :param directory_path: The directory to remove
-    :type directory_path: str
-    """
-    try:
-        shutil.rmtree(directory_path)
-    except OSError as error:
-        # if perm denied (13) add rwx permissions to all directories and retry
-        # so that we are not blocked by users creating directories with no permissions
-        if error.errno is errno.EACCES:
-            for root, dirs, files in os.walk(directory_path):
-                for dir_path in dirs:
-                    os.chmod(os.path.join(root, dir_path),
-                             stat.S_IXUSR | stat.S_IWUSR | stat.S_IREAD)
-            shutil.rmtree(directory_path)
-        else:
-            raise
-
-
-@contextmanager
-def working_dir_context():
-    """
-    Prepares a working directory that is removed after the context has ended.
-    """
-    try:
-        working_dir = get_working_directory()
-        os.chdir(working_dir)
-        yield working_dir
-    finally:
-        delete_working_directory()
+            return os.path.join(
+                root,
+                task.current.request.hostname,
+                task.current.request.id)
+        except AttributeError:
+            raise RuntimeError(_('WorkingDirectory may only be used within a Task.'))
+
+    def __init__(self):
+        """
+        Raises:
+            RuntimeError: When used outside a Celery task.
+        """
+        self._path = self._build_path()
+
+    @property
+    def path(self):
+        """
+        Returns:
+            str: The absolute directory path.
+        """
+        return self._path
+
+    def create(self):
+        """
+        Create the directory.
+        Deleted and recreated when already exist.
+        """
+        try:
+            os.makedirs(self.path)
+        except FileExistsError:
+            self.delete()
+            os.makedirs(self.path)
+
+    def delete(self):
+        """
+        Delete the directory (tree).
+        On permission denied - an attempt is made to recursively fix the
+        permissions on the tree and the delete is retried.
+        """
+        try:
+            shutil.rmtree(self.path)
+        except PermissionError:
+            for path in os.walk(self.path):
+                os.chmod(path[0], stat.S_IXUSR | stat.S_IWUSR | stat.S_IREAD)
+            shutil.rmtree(self.path)
+
+    def __enter__(self):
+        """
+        Create the directory and set the CWD to the path.
+
+        Returns:
+            WorkingDirectory: self
+
+        Raises:
+            OSError: On failure.
+        """
+        self.create()
+        os.chdir(self.path)
+        return self
+
+    def __exit__(self, *ignored):
```

Comment:
IIRC, this will cause any exceptions to be swallowed. I think we want to reraise.

Reply:
Exception raised within the content do not need to be raised in the __exit__() to prevent swallowing.  The arguments passed to __exit__() are so that the code can know an exception was raised within the content so it can make clean up decisions.  No to swallow them.